content/5_Plat/Binary_Jump.mdx

```diff
@@ -31,77 +35,117 @@ prerequisites:
 	/>
 </Resources>
 
-<Info title="Pro Tip">
-
-Binary jumping is more commonly referred to as "binary lifting."
+### Explanation
 
-</Info>
-
-### Solution
-
-To solve this problem, we can use a standard binary lifting implementation where
-`jmp(int x, int d)` corresponds to the $d$-th ancestor of $x$.
+Binary lifting consists of calculating the $2^k$-th ancestor of each node
+for all relevant values of $k$ and storing them in a table.
 
-In our `jmp(int x, int d)` if our final value of $x$ is $0$, then such a node
-does not exist and we can simply return $-1$. This is because the lowest number
-a node can be is $1$ (the root of the tree).
+With this table, we can then efficiently answer queries regarding
+the $k$-th ancestor of all nodes.
+This is because any $k$ can be broken down into a sum of powers of $2$
+with its binary representation.
 
-In our implementation, we test if we jump in powers of two by using the $\&$
-operator. If the $i$th bit on the right is toggled, then we jump. For example, a
-jump of $13$ would correspond to the binary number $1101$. We would jump 3 times
-on bits $0, 2, 3$ (in that order) counting from the right.
+This way, instead of directly computing, say, the $13$th ancestor of a node,
+we can go to the $8$th, then $4$th, then $1$st ancestor of the node.
+This results in a logarithmic complexity for computing the $k$th parent.
 
-<Spoiler title="Illustration of the jump method">
+Here's an animation of how we jump if you're still confused:
 
-![Binary Lifting Diagram](./assets/binaryliftingjump.png)
+<video width="960" height="720" controls>
+	<source src="/animations/binary_jumping.mp4" type="video/mp4"/>
+</video>
 
-</Spoiler>
+To actually compute our binary jumping table, we start with the $2^0=1$st
+parents of each node, which is their direct parent.
+We then move on and calculate $2^1=2$nd parents, using the fact that
+the $2$nd parent can be calculated as the $1$st parent of the $1$st parent.
+Using similar logic, we move on to $2^2$, $2^3$, and so on.
+We stop when $2^n$ is greater than the size of the tree, since at that point
```

Comment:
i think we should keep the part where it talks about how to calculate the number of rows for the binlift table

Reply:
> We stop when $2^n$ is greater than the size of the tree

unless you think this isn't enough?

Reply:
fair enough 

also another thing, setting the number of rows to ceil(log2(n)) is only enough for 2^n - 1, which doesn’t matter for LCA but can matter for using binlift for range queries

Reply:
why's that? if it goes up to 2^n, then you can have at most 2^n parents

Reply:
it doesn’t matter cuz for most tree problems you only need to lift up n-1 times (eg if the tree is a line). but if you are using binlift for range queries, i’m p sure ceil(log2(n)) is problematic. i’ll try to find an example later (or i’m just tripping)

Reply:
so uh
did you find a sussy test case yet

Reply:
an example of when ceil(log2(n)) does not work for range queries:

ac: https://codeforces.com/contest/1142/submission/261280040
rte: https://codeforces.com/contest/1142/submission/265726905

granted, the module primarily talks about binlift for tree/graph stuff. but I think using something like 1 + (int) log2(n) is a better way to write it in general (also because you can write it as 1 + __lg(n), which doesn't deal with decimal precision stuff)

Reply:
wait no now i'm curious about why this is the case, think you could explain why?

Reply:
> wait no now i'm curious about why this is the case, think you could explain why?

ok actually the submissions i linked are a BIT of a bad example. in this case, it RTEs because ceil(log2(n)) evaluates to 0 when n = 1, so accessing the binlift array in any capacity RTEs cuz there aren't any rows...

a better example of when this would WA is attached [here](https://ide.usaco.guide/O-L4nyUb3Rzd1lL1q3S) (you can obv write ur code so that ceil(log2(n)) would work, but it's easier to just use 1 + (int)log2(n) or smth like that)

below is the write up i did for why it can be wrong:
<details><summary>

</summary>

Let $x$ be the number of rows in the table. Because the highest bit it covers is $x - 1$, at most we can lift $2^{x} - 1$ times. Now, say we set $x$ to be ceil(log2(n)). If n is a perfect power of 2, then $2^x $is equal to exactly $n$. So, this means that we can only lift $n-1$ times. However, if we need to lift $n$ times (e.g. in a range query), this is problematic.

</details>

Reply:
then why doesn't it happen when doing ancestor queries

Reply:
ok so i figured it out- this current method works for just lca so i'll leave it be